org.eclipse.xtext.xbase/src/org/eclipse/xtext/xbase/typesystem/conformance/TypeConformanceComputer.java

```diff
@@ -206,107 +207,419 @@ protected int doIsConformant(FunctionTypeReference left, FunctionTypeReference r
+			}
+			if (superType instanceof UnboundTypeReference)
+				return superType;
+			if (allBoundNames.size() != allNames.size()) {
+				if (allBoundNames.size() == 1 && allBoundNames.contains(getIdentifier(superType))) {
+					return superType;
+				}
 			}
+			WildcardTypeReference result = owner.newWildcardTypeReference();
+			result.addUpperBound(superType.copyInto(owner));
 			return result;
 		}
-		return null;
+		
+		protected boolean isRecursiveRequest(List<LightweightTypeReference> types, Set<String> allNames) {
+			for(List<LightweightTypeReference> previousRequest: requestsInProgress) {
+				if (doIsRecursiveRequest(types, allNames, previousRequest)) {
+					return true;
+				}
+			}
+			return false;
+		}
+
+		protected boolean doIsRecursiveRequest(List<LightweightTypeReference> types, Set<String> allNames, List<LightweightTypeReference> initiallyRequested) {
+			if (types.size() < initiallyRequested.size()) {
+				return false;
+			}
+			for(LightweightTypeReference initialRequest: initiallyRequested) {
+				if (!allNames.contains(getIdentifier(initialRequest))) {
+					return false;
+				}
+			}
+			return true;
+		}
+		
+		private void addIdentifier(LightweightTypeReference type, Set<String> allNames, Set<String> allBoundNames) {
+			if (type instanceof UnboundTypeReference && !type.isResolved()) {
+				allNames.add(((UnboundTypeReference) type).getHandle().toString());
+			} else {
+				String identifier = type.getJavaIdentifier();
+				allNames.add(identifier);
+				allBoundNames.add(identifier);
+			}
+		}
+		
+		private String getIdentifier(LightweightTypeReference type) {
+			if (type instanceof UnboundTypeReference && !type.isResolved())
+				return ((UnboundTypeReference) type).getHandle().toString();
+			return type.getJavaIdentifier();
+		}
+		
+		private LightweightTypeReference getMostSpecialTypeIfAllWildcards(List<LightweightTypeReference> types) {
+			boolean objectIsCandidate = false;
+			boolean lowerBoundSeen = false;
+			for(LightweightTypeReference type: types) {
+				if (type instanceof WildcardTypeReference) {
+					if (((WildcardTypeReference) type).getLowerBound() == null) {
+						objectIsCandidate = true;
+					} else {
+						lowerBoundSeen = true;
+					}
+				} else {
+					return null;
+				}
+			}
+			if (!lowerBoundSeen) {
+				return null;
+			}
+			if (objectIsCandidate)
+				return owner.newWildcardExtendsObject();
+			return getMostSpecialType(types);
+		}
+
+	}
+	
+	/**
+	 * Compute the common super type for the given types.
+	 * 
+	 * May return <code>null</code> in case one of the types is primitive void but not all 
+	 * of them are.
+	 */
+	/* @Nullable */
+	public LightweightTypeReference getCommonSuperType(final /* @NonNull */ List<LightweightTypeReference> types, ITypeReferenceOwner owner) {
+		return newCommonSuperTypeFinder(owner).getCommonSuperType(types);
+	}
+
+	private CommonSuperTypeFinder newCommonSuperTypeFinder(ITypeReferenceOwner owner) {
```

Comment:
maybe we should make this protected in case someone wants to change

Reply:
Yes, that was my intention, but I forgot to change the visibility after 'Extract Method'. Done.